### PR TITLE
Feature/4840 show suppliers returned

### PR DESCRIPTION
**Pull request by alibombaye**
Story [AB#4840](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/4840)
Task [AB#7102](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/7102)

Just a note I had to move the actual finding of suppliers to the /select route. Since we won't be able to share the suppliers found from /search to /select without doing a search again in /select.

Now the /search just validates that something is entered and if so redirects to /select with the query string appended of what was entered.

Screenshot 2020-05-22 at 10 24 20

### CODE REVIEW

**Comment by alibombaye** [comment]
We may want to change the route of the /select to /search/:suppliersearchstring
rather than pass if through via a query string.

I am thinking of when we want to go back from the supplier section page to the select page. We will need this suppliersearchstring again to do the find.

Or do we think now is a good time to start thinking about using redis state?

File: app/pages/sections/supplier/select/contextCreator.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import manifest from './manifest.json';
+import { baseUrl } from '../../../../config';
+
+const generateSupplierOptions = ({ suppliers }) => {
+  const supplierOptions = suppliers.map(supplier => ({
```
Comment (kat1906) [inline app/pages/sections/supplier/select/contextCreator.js:5]: can just return from this line:
const generateSupplierOptions = ({ suppliers }) => suppliers.map(supplier => ({
    value: supplier.supplierId,
    text: supplier.name,
  }));

File: app/pages/sections/supplier/select/contextCreator.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    value: supplier.supplierId,
+    text: supplier.name,
+  }));
+
+  return supplierOptions;
+};
+
+const generateQuestionsContext = ({ suppliers }) => {
+  const questions = manifest.questions.map(question => ({
```
Comment (kat1906) [inline app/pages/sections/supplier/select/contextCreator.js:14]: same here

File: app/pages/sections/supplier/select/contextCreator.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    ...question,
+    options: generateSupplierOptions({ suppliers }),
+  }));
+
+  return questions;
+};
+
+export const getContext = ({ orderId, suppliers }) => {
+  const context = ({
```
Comment (kat1906) [inline app/pages/sections/supplier/select/contextCreator.js:23]: and here

File: app/pages/sections/supplier/select/controller.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+export const findSuppliers = async ({ supplierNameToFind, accessToken }) => {
+  const endpoint = getEndpoint({ endpointLocator: 'getSearchSuppliers', options: { supplierNameToFind } });
+  const suppliersFound = await getData({ endpoint, accessToken, logger });
+  logger.info(`Searching for "${supplierNameToFind}" returned ${suppliersFound.length} supplier`);
+
+  return suppliersFound;
+};
+
+export const getSupplierSelectPageContext = ({ orderId, suppliers }) => (
```
Comment (kat1906) [inline app/pages/sections/supplier/select/controller.js:14]: can shorten this to:
export const getSupplierSelectPageContext = (params) => getContext(params)

File: app/pages/sections/supplier/select/ui.test.js [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    .expect(getLocation()).eql('http://identity-server/login');
+});
+
+test('should render Supplier select page', async (t) => {
+  nock(solutionsApiUrl)
+    .get('/api/v1/suppliers?name=some-supp')
+    .reply(200, [{}]);
+
+  await pageSetup(t, true);
+  await t.navigateTo(`${pageUrl}?supplierNameToFind=some-supp`);
```
Comment (kat1906) [inline app/pages/sections/supplier/select/ui.test.js:55]: really don't like the query - just use name? it is not like we are using any others here

Comment (kat1906) [reply app/pages/sections/supplier/select/ui.test.js:55]: this part is user facing after all

File: app/pages/sections/supplier/select/ui.test.js [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    .expect(page.exists).ok();
+});
+
+test('should navigate to /organisation/order-1/supplier/search when click on backlink', async (t) => {
+  nock(solutionsApiUrl)
+    .get('/api/v1/suppliers?name=some-supp')
+    .reply(200, [{}]);
+
+  await pageSetup(t, true);
+  await t.navigateTo(`${pageUrl}?supplierNameToFind=some-supp`);
```
Comment (kat1906) [inline app/pages/sections/supplier/select/ui.test.js:68]: you have the pageUrl variable at the top, you could add another with the query string?

File: app/routes.js [hunk 7]
```diff
@@ -137,12 +152,7 @@ export const routes = (authProvider) => {
       });
     }
 
-    const context = await getSupplierSearchPageErrorContext({
-      orderId,
-      validationErrors: response.errors,
-    });
-
-    return res.render('pages/sections/supplier/search/template.njk', addContext({ context, user: req.user, csrfToken: req.csrfToken() }));
+    return res.redirect(`${config.baseUrl}/organisation/${orderId}/supplier/search`);
```
Comment (kat1906) [inline app/routes.js:155]: where is the redirect to the 0 found page?

File: app/pages/sections/supplier/select/ui.test.js (hunk 5 above)
Comment (alibombaye) [reply app/pages/sections/supplier/select/ui.test.js:55]: This will be removed once we have state. OK to leave as is for now?

File: app/routes.js (hunk 7 above)
Comment (alibombaye) [reply app/routes.js:155]: the throw new ErrorContext on like 146

File: app/pages/sections/supplier/select/ui.test.js (hunk 5 above)
Comment (kat1906) [reply app/pages/sections/supplier/select/ui.test.js:55]: I would like it changed (just in case) - you gan do global search and replace - will take 30secs

File: app/pages/sections/supplier/select/ui.test.js (hunk 6 above)
Comment (alibombaye) [reply app/pages/sections/supplier/select/ui.test.js:68]: again will be removed once we have state. OK to leave for now?

File: app/pages/sections/supplier/select/ui.test.js (hunk 5 above)
Comment (alibombaye) [reply app/pages/sections/supplier/select/ui.test.js:55]: haha ok fair enough. Like the vote of confidence 😢

File: app/pages/sections/supplier/select/ui.test.js (hunk 6 above)
Comment (kat1906) [reply app/pages/sections/supplier/select/ui.test.js:68]: global search and replace 😆
